### ISSUE TITLE
Word embedding (predicting next word)

### ISSUE THREAD

**Issue by worldofpiggy** [fchollet/keras#2009]
I am training a network to predict the next  word from a context window of maxlen words.
I feed the network with a pair (x,y) where 
x is a list of maxlen word indices and 
y is the index of the next word

I need to learn the embedding of all vocsize words 
Here is the model:

```
 model = Sequential()
layers = [maxlen,256,256,1]
model.add(Embedding(vocsize, 128))

model.add(LSTM(input_dim=layers[0], output_dim=layers[1], return_sequences=True))
model.add(Dropout(0.2))

model.add(LSTM(layers[2], return_sequences=False))
model.add(Dropout(0.2))

model.add(Dense(output_dim = layers[3]))
model.add(Activation("linear"))
#model.add(Activation('sigmoid'))
model.compile(loss="mse", optimizer="rmsprop")
#model.compile(loss='binary_crossentropy', optimizer='rmsprop')
```

When I fit it to x and y I get a loss of -5444.4293 steady for all epochs. It doesn't seem to learn anything.
What am I doing wrong?

**Comment by ymcui** [comment]
Y should be in shape of (batch_size, vocab_size), instead of (batch_size, 1).
say, the Y should be in one-hot representations, not word indices.

**Comment by worldofpiggy** [comment]
thanks a lot ymcui. 
I'm running on some epochs and see what I get...

**Comment by worldofpiggy** [comment]
I concatenated the text of three books, to get about 20k words and enough text to train.
I cut sentences of 10 words and want to predict the next word after 10. Hence, I am feeding the network with 10 word indices (into the Embedding layer) and a boolean vector of size <vocsize> for the next word to predict.
I am also using sigmoid and rmsprop optimizer.
After 150 epochs I get no more improvement on the loss and if I plot the Embedding with t-sne there is basically no structure in the similarity of the words... nor syntax nor semantics...

Any hint?

Here is the model

maxlen = 10
model = Sequential()
layers = [maxlen, 256, 512, vocsize]
model.add(Embedding(vocsize, 300))
model.add(LSTM(input_dim=layers[0], output_dim=layers[1], return_sequences=False)) 
model.add(Dropout(0.5))   
model.add(Dense(output_dim = layers[3])) 
model.add(Activation('sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='rmsprop')

**Comment by ymcui** [comment]
Here are some suggestions:
1) replace final `sigmoid` with `softmax`, because in language modeling, the final output should be a probability distribution, which makes `softmax` activation more make sense than `sigmoid`.
2) try using `categorical_crossentropy` objectives.

**Comment by worldofpiggy** [comment]
Executing.
Now the loss makes much more sense across epochs. It started from 6.9 and is going down as I've seen it in working networks, ~0.12 per epoch. Will keep you posted.
Thanks for the hint!

**Comment by Pratapdangeti** [comment]
@worldofpiggy I too looking for similar solution, could you please share me complete code ? Thanks in advance

**Comment by AdamKing11** [comment]
Hey y'all,
This is about a year later, but I think I may know why you're having your NN never gain any accuracy. I was trying to do a very similar thing with the Brown corpus - use word embeddings rather than one-hot vector encoding for words to make a predictive LSTM - and I ran into the same problem.
After sitting and thinking for a while, I think the problem lies in the output and the output dimensions. Right now, your output 'y' is a single scalar, the index of the word, right? As you have it in your last post, the output layer will shoot out a vocabulary-sized vector of real-valued numbers between 0 and 1. You'll probably be able to get it to work if you instead convert the output to a one-hot representation of its index.

Of course, I'm still a bit of a newbie in Keras and NN's in general so think might be totally way off....

tl;dr: Try making your outputs one-hot vectors, rather that single scalar indexes.

**Comment by kaushalshetty** [comment]
Hi @worldofpiggy 
Take the whole text data in a string and tokenize it using keras.preprocessing.text. Then take a window of your choice say 100. Create a new training data set  each of 100 words and (100+1)th word becomes your label. 
x = [ [hi,how,are,......], [is,that,on,say,.....], [ok,i,am,is.....]]
y = [is,ok,done]
Now combine x into sentences like :   
x = [hi how are ......    ,  is that on say ...   ,  ok i am is .....]   #this step is done to use keras tokenizer
Now use keras tokenizer to tokenize them and do a text to sequence to it
x = [[1,2,3,....] , [4,56,2 ...] , [3,4,6 ...]]
Map y to tokenizer.word_index and convert it into a categorical variable .
y = [10,11,12]
convert  x into numpy and reshape it into (train_data_size,100,1)
Fit the lstm model
Obtain the index of y having highest probability.
Reverse map this using the word_index



You can repeat this for any number of sequences.
Good Luck!

**Comment by kaushalshetty** [comment]
Also use categorical_crossentropy  and softmax in your code

**Comment by rahulkulhalli** [comment]
Hey @kaushalshetty,

Could you please elaborate the procedure? It'd be really helpful. Thanks!
